### PR TITLE
Teams backend

### LINKED ISSUE #61: Add edition teams backend
In order to facilitate moderating and proper maintenance of other parts of the website we will use the following approach:
Add a way for students/users to be "promoted" to "staff" as part of a department of the SINF team for that edition, these users are able to access some management options, such as validating participant codes in talks/workshops, and will be displayed in the teams page (front end done in #55 ).
This could be done by adding a department option to the existent student.

### PR DESCRIPTION
**Pull request by toni-santos**
Closes #61

### CODE REVIEW

File: database/migrations/2023_09_15_133618_create_departments_table.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        });
+
+    }
+
+    /**
+     * Reverse the migrations.
+     */
+    public function down(): void
+    {
+        Schema::dropIfExists('teams');
```
Comment (ttoino) [inline database/migrations/2023_09_15_133618_create_departments_table.php]: You should drop departments and staff, not this table.

File: app/Models/Student.php [hunk 2]
```diff
@@ -48,4 +49,9 @@ public function user(): BelongsTo
     {
         return $this->belongsTo(User::class);
     }
+
+    public function staff(): BelongsToMany
+    {
+        return $this->belongsToMany(Staff::class);
+    }
```
Comment (ttoino) [inline app/Models/Student.php]: I think this should be a hasMany relationship

File: app/Models/Staff.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    protected $fillable = [
+        'coordinator',
+        'department_id',
+        'student_id',
+    ];
+
+    public function student(): HasMany
+    {
+        return $this->hasMany(Student::class);
+    }
```
Comment (ttoino) [inline app/Models/Staff.php]: This should be a belongsTo relationship

File: app/Http/Controllers/StaffCRUDController.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+class StaffCRUDController extends CRUDController
+{
+    protected string $model = Staff::class;
+
+    protected string $view = 'Staff';
+
+    protected array $rules = [
+        'student_id' => 'required|integer',
+        'department_id' => 'required|integer',
```
Comment (ttoino) [inline app/Http/Controllers/StaffCRUDController.php]: These should validate against the relevant tables. I can't remember the syntax off the top of my head but it's something like `exists:students,id`

File: app/Http/Controllers/StaffCRUDController.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    protected function updated(array $old, array $new): ?array
+    {
+        return [
+            'coordinator' => $new['coordinator'],
+            'department_id' => $new['department_id'],
+            'student_id' => $new['student_id'],
+        ];
+    }
```
Comment (ttoino) [inline app/Http/Controllers/StaffCRUDController.php]: These functions are not needed if you're not doing something more advanced

File: app/Http/Controllers/DepartmentCRUDController.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        ];
+    }
+
+    protected function updated(array $old, array $new): ?array
+    {
+        return [
+            'edition_id' => $new['edition_id'],
+            'name' => $new['name'],
+        ];
+    }
```
Comment (ttoino) [inline app/Http/Controllers/DepartmentCRUDController.php]: Same as above

File: resources/js/Pages/CRUD/Department/Create.vue [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                label="Nome"
+                type="text"
+                required
+                autofocus
+                autocomplete="name"
+                :error-message="form.errors.name"
+            />
+
+            <PrimaryButton type="submit">Criar</PrimaryButton>
+        </form>
```
Comment (ttoino) [inline resources/js/Pages/CRUD/Department/Create.vue]: I'd prefer name to be first, but it's just a nitpick

File: resources/js/Pages/CRUD/Department/Edit.vue [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                type="text"
+                required
+                autofocus
+                autocomplete="name"
+                :error-message="form.errors.name"
+            >
+            </TextInput>
+
+            <PrimaryButton type="submit">Editar</PrimaryButton>
+        </form>
```
Comment (ttoino) [inline resources/js/Pages/CRUD/Department/Edit.vue]: Same as above

File: resources/js/Types/Department.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+import type Model from "./Model";
+
+export default interface Department extends Model {
+    id: number;
```
Comment (ttoino) [inline resources/js/Types/Department.ts]: `id` is redundant here, `Model` already provides it

File: resources/js/Types/Staff.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+import type Model from "./Model";
+
+export default interface Department extends Model {
+    id: number;
```
Comment (ttoino) [inline resources/js/Types/Staff.ts]: Same as above

**Review by ttoino — CHANGES_REQUESTED** [review]

File: app/Models/Edition.php [hunk 11]
```diff
@@ -40,6 +40,11 @@ public function quests(): HasMany
         return $this->hasMany(Quest::class);
     }
 
+    public function team(): HasMany
```
Comment (ttoino) [inline app/Models/Edition.php]: This should be `departments`, and use the `Department` class

File: app/Models/Staff.php [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        'department_id',
+        'student_id',
+    ];
+
+    public function student(): BelongsTo
+    {
+        return $this->belongsTo(Student::class);
+    }
+
+    public function team(): BelongsTo
```
Comment (ttoino) [inline app/Models/Staff.php]: This should be `department`

File: resources/js/Pages/CRUD/Staff/Create.vue [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+            </select>
+
+            <select v-model="form.student_id" required class="self-stretch">
+                <option value="" disabled selected hidden>User</option>
+                <option
+                    v-for="student in $props.with.students"
+                    :key="student.id"
+                    :value="student.id"
+                >
+                    {{ `${student.user?.name} - ${student.user?.id}` }}
```
Comment (ttoino) [inline resources/js/Pages/CRUD/Staff/Create.vue]: You don't need to use string interpolation here

```suggestion
                    {{ student.user?.name }} - {{ student.user?.id }}
```

File: resources/js/Pages/CRUD/Staff/Edit.vue [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+            </select>
+
+            <select v-model="form.student_id" required class="self-stretch">
+                <option value="" disabled selected hidden>User</option>
+                <option
+                    v-for="student in $props.with.students"
+                    :key="student.id"
+                    :value="student.id"
+                >
+                    {{ `${student.user?.name} - ${student.user?.id}` }}
```
Comment (ttoino) [inline resources/js/Pages/CRUD/Staff/Edit.vue]: Same as above

**Review by ttoino — CHANGES_REQUESTED** [review]

File: app/Models/Department.php [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+class Department extends Model
+{
+    use HasFactory;
+
+    protected $fillable = [
+        'name',
+        'edition_id',
+    ];
+
+    public function department(): BelongsTo
```
Comment (Naapperas) [inline app/Models/Department.php]:
```suggestion
    public function edition(): BelongsTo
```

**Review by Naapperas — CHANGES_REQUESTED** [review]

File: resources/js/Pages/CRUD/Staff/Create.vue [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+                <option
+                    v-for="department in $props.with.departments"
+                    :key="department.id"
+                    :value="department.id"
+                >
+                    {{
+                        `${department.name} - ${
+                            editions[department.edition_id]
+                        }`
+                    }}
```
Comment (ttoino) [inline resources/js/Pages/CRUD/Staff/Create.vue]: You don't need string interpolation here

File: resources/js/Pages/CRUD/Staff/Edit.vue [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+                <option
+                    v-for="department in $props.with.departments"
+                    :key="department.id"
+                    :value="department.id"
+                >
+                    {{
+                        `${department.name} - ${
+                            editions[department.edition_id]
+                        }`
+                    }}
```
Comment (ttoino) [inline resources/js/Pages/CRUD/Staff/Edit.vue]: Same as above

File: app/Models/Edition.php [hunk 18]
```diff
@@ -44,4 +44,9 @@ public function sponsors(): HasMany
     {
         return $this->hasMany(Sponsor::class);
     }
+
+    public function team(): HasMany
+    {
+        return $this->hasMany(Team::class);
+    }
```
Comment (ttoino) [inline app/Models/Edition.php]: Same as last review

**Review by ttoino — CHANGES_REQUESTED** [review]

**Review by ttoino — DISMISSED** [review]

**Review by Naapperas — DISMISSED** [review]

**Review by Naapperas — APPROVED** [review]

**Review by ttoino — APPROVED** [review]

**Pull request merged**